Approved/dyn-type.md

```diff
@@ -0,0 +1,196 @@
+
+### Dynamic values require runtime support
+
+The example generates an error when calling `Foo` with a dynamic value as argument `i` unless the target supports arrays.
+
+The developer can create values at runtime by measuring a qubit and then pass these values to other callables that process it further and return it.
+
+> Define the array index as a callable input parameter
+
+```qsharp
+operation Foo(b : Bool, i : Int) : Int {
+    return [1, 2, 3][i]; // This needs runtime support if i is dynamic
+}
+
+operation Bar() : Int {
+// get a dynamic value (usually by measuring a qubit)
+    use q = Qubit();
+    H(q);
+    let val = M(q) == Zero ? 0 : 1; // val is of type dyn Int
+
+    let res = Foo(true, val); // requires that the runtime supports arrays
+    return res; // res is of type dyn Int
+}
+```
+
+> Define the lifting operator for the array index
+
+```qsharp
+lift_i Foo : (Bool, dyn Int) => dyn Int
+```
+
+> The array index is lifted into dyn. The result value of the operation is also a dyn.
+
+
+### Lift a boolean into dyn
+
+> Define the boolean as a callable input parameter
+
+```qsharp
+operation Foo(b : Bool, i : Int) : Int {
+    use q = Qubit();
+    if b { X(q); }
+    return [1, 2, 3][i];
+}
+```
+
+> Define the lifting operator for the boolean
+
+```qsharp
+lift_b Foo : (dyn Bool, Int) => Int
+```
+
+> The boolean is lifted into `dyn`. The result value of the operation is not `dyn`. The result is opaque as `b` is used to branch for a gate.
+
+### Cannot lift anonymous parameters into dyn
+
+> Define anonymous parameters as a callable input parameter
+
+```qsharp
+operation Foo(Bool, Int) : Int {
+    use q = Qubit();
+    if b { X(q); }
+        return [1, 2, 3][i];
+}
+```
+
+> It is not possible to define the lifting operator for the anonymous
+
+```qsharp
+(Bool, Int) => Int
+```        
+> The anonymous parameter is not lifted into dyn
+
+### Compiler injects values into dyn
+
+> The developer sets up a conditional statement for the injection
+
+```qsharp
+If x : T, then inject x : dyn T
+```
+
+> The compiler automatically injects compile-time values into dyn
+
+## Implementation
+
+The implementation will be informed by below design goals:
+> Goal 1: Keep compilation and optimization logic within llvm as much as possible, rather than replicating it in the Q# compiler
+
+> Goal 2: Q# functions serve the same purpose as const functions in other languages, i.e. we basically want to guarantee that a function can be evaluated when and as soon as all arguments are known. This in particular means that unless an argument is `dyn`, we expect that the function call will be fully evaluated/evaluatable at compile time.
```

Comment:
Does that imply that a function can not return a `dyn` unless at least one input was `dyn`?